src/librustc_lint/levels.rs

```diff
@@ -95,6 +96,44 @@ impl<'s> LintLevelsBuilder<'s> {
         self.sets.list.push(LintSet::CommandLine { specs });
     }
 
+    /// Attempts to insert the `id` to `level_src` map entry. If unsuccessful
+    /// (e.g. if a forbid was already inserted on the same scope), then emits a
+    /// diagnostic with no change to `specs`.
+    fn insert_spec(
+        &mut self,
+        specs: &mut FxHashMap<LintId, LevelSource>,
+        id: LintId,
+        (level, src): LevelSource,
+    ) {
+        if let Some((old_level, old_src)) = specs.get(&id) {
+            if old_level == &Level::Forbid && level != Level::Forbid {
+                let mut diag_builder = struct_span_err!(
+                    self.sess,
+                    src.span(),
+                    E0453,
+                    "{}({}) incompatible with previous forbid in same scope",
```

Comment:
Interesting, so it is a hard error now to have `forbid` and something else on the same scope, but not in a nested scope?

I guess I don't care very strongly, just didn't expect this. I expected the same behavior was for nested scopes. Also I think this is inconsistent with command-line handling (where `-F lint -A lint` works, but the `-A` has no effect).

Reply:
I don't understand your response.

Its a hard error to have a forbid on one scope and something else in a nested scope.  (AIUI, that's the whole point of forbid, versus deny).

For example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=63cb51c86b8b1c15a8b22f132ce6c32e

Can you show me an example of what current behavior of forbid that you think I should be modelling instead?

Reply:
Oops, somehow I thought `forbid` would *overwrite* later `allow`, I did not know that it makes them an error. Never mind then. I think my only comment for this PR then is that it would be nice to share the code that emits that error -- if the error is pre-existing; the wording is slightly different now.

I feel command-line arguments should be consistent with lint attributes though.

Reply:
The problem with making command line arguments "consistent" with lint attributes, from my point of view, is that I think a lot of tools do inject (or freely mix) new command line arguments, and so I think users run into combinations of `-F` and `-A` on a single command line a lot more often than they run into `#[forbid]` and `#[allow]` in a single source item. 

A hard error for mixing `-F` and `-A` when they arose due to e.g. makefile variables is really annoying for end users.

(I suppose the analogy for source code items would be macro expansion, right? Does one see `#[allow]` getting injected from macros very often? I can admit this has crossed my mind at times, especially back when I put in the `push_unsafe`/`pop_unsafe` hacks [five years ago](https://github.com/rust-lang/rust/pull/27215/commits/1829fa5199bae5a192c771807c532badce14be37) ...)

Reply:
Indeed I'd argue that for the same reasons, a hard error could be annoying for the attributes. But then nobody complaining about it for five years is a good argument for it not being very annoying.^^